File: 0938. Range Sum of BST.java
// Easy
// DFS
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/range-sum-of-bst/

class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        return dfs(root, low, high);
    }
    private int dfs(TreeNode node, int low, int high) {
        if (node == null) {
            return 0;
        }
        int sum = 0;
        if (node.val >= low && node.val <= high) {
            sum += node.val;
        }
        if (node.val > low) {
            sum += dfs(node.left, low, high);
        }
        if (node.val < high) {
            sum += dfs(node.right, low, high);
        }
        return sum;
    }
}

File: 0501-1000/0938. Range Sum of BST.py
# Time:O(n), Space:O(h)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        self.ans = 0

        def inorder(node):
            if not node:
                return
            
            inorder(node.left)
            if low <= node.val <= high:
                self.ans += node.val
            inorder(node.right)

        inorder(root)
        return self.ans
    